src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -92,6 +97,32 @@ private static bool IsValidForTaskLikeTypeOnlyContext(ISymbol symbol, TSyntaxCon
             return symbol.IsAwaitableNonDynamic(context.SemanticModel, context.Position);
         }
 
+        private static bool IsValidForGenericConstraintContext(ISymbol symbol, TSyntaxContext context)
+        {
+            if (symbol.IsNamespace() ||
+                symbol.IsKind(SymbolKind.TypeParameter))
+            {
+                return true;
+            }
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType() ||
+                namedType.IsStructType())
+            {
+                return false;
+            }
+
+            // If current symbol is static or sealed then it cannot be used as a generic constraint.
+            // However it can contain other valid constraint types and if this is true we should show it
+            if (namedType.IsStatic || namedType.IsSealed)
+            {
+                return context.SemanticModel.LookupSymbols(context.Position, namedType).Any(s => IsValidForGenericConstraintContext(s, context));
```

Comment:
I'm curious about the potential cost of binding all the members for each type symbols here. I understand this is rarely exercised (only in generic constraint context,) but I wouldn't mind having those possibly invalid types in the list as long as the ones I need are in there too.

Reply:
For https://github.com/dotnet/roslyn/pull/60528 we checked `IsAwaitableNonDynamic`, which does the same thing when searching for `GetAwaiter` method, for _every_ incoming class. Now when VS 17.3 P1 is an available public release and this feature is in it, from my experience delay appears only in specific conditions, like if I start typing while VS is loading the solution, and this delay is always under a second. Here amount of such operations is considerably smaller, so I don't expect any perf issues

Reply:
idon't think this is accurate.  effectively, we just want to check if there's a nested type within.  i don't think we need LookupSymbols.  Just GetTypeMembers would likely suffice.  I do worry about this though as it will cause us to have to load metadata for all nested types within a metadata symbol and the like.

Reply:
@CyrusNajmabadi Yes, `GetTypeMembers` is definitely better approach here! About loading metadata for nested types: relatively to already loaded metadata for non-nested types, the amount of metadata for nested types is quite small. Nested types are not that common pattern, so I don't think we should wory about it.

Reply:
The issue is not necessarily even the data for those nested types.  It's that we do have to do all this population of information period.  e.g. every sealed/static symbol (including in all the types we import from everywhere in the universe) will have to allocate the information (even to store the empty state).

It may be ok.  But it's something we need to be cautious about.  

For context, we did this back in the native days.  And we had to roll it back due to the extra costs.  It may be though that enough time has passed that now makes this acceptable.

@genlu for a large project with a lot of references, can you collect how many types fall into this category (including when "show non-imported types" is on)?    Thanks!

Reply:
> for a large project with a lot of references, can you collect how many types fall into this category

Are you asking for the number of types contain nested types in a solution like Roslyn?

> including when "show non-imported types" is on

Sure, but just to be clear, this PR doesn't do any filtering for unimported types.

Reply:
ah, that's a good point.  that makes this feel a lot safer then.